### ISSUE TITLE
listOf ObjectType with Deferred resolve

### ISSUE THREAD

**Issue by Tam** [webonyx/graphql-php#382]
I'm trying to write an api that will build an SQL query as the graph query is parsed and resolve all the data once the query is complete, allowing me to cherry-pick the columns I need across multiple database tables. I've had some success when only querying one element, but trying to return more than one by adding `listOf` into the mix causes `expected iterable, but did not find one` to be thrown.

```php
class Types
{
    public function query ()
    {
        return new QueryType([
            'name' => 'Query',
            'fields' => [
                'events' => Type::listOf($this->get('events')), // <- Fails
                // 'events' => $this->get('events'), // <- Works
            ],
            'resolveField' => function ($source, $args, $context, ResolveInfo $info) {
                return $this->{$info->fieldName}($source, $args, $context, $info);
            },
        ]);
    }

    public function events ()
    {
        return new QueryType([
            'name' => 'Events',
            'fields' => [
                'id' => Type::id(),
            ],
            'resolveField' => function ($source, $args, $context, ResolveInfo $info) {
                $field = $info->fieldName;

                if (method_exists($this, $field))
                    return $this->{$field}($source, $args, $context, $info);

                Buffer::add('events', $field);

                return new Deferred(function () use ($field) {
                    Buffer::load();
                    return Buffer::get('events', $field);
                });
            },
        ]);
    }
}
```

I need to have the entire graph query parsed and build into an SQL query before any data is resolved. Judging by the docs, examples, and my own testing `listOf` requires that the data be returned immediately, meaning that resolving `new Deferred` never works. This in turn means I can't cherry-pick the data I need, rather that I need to get everything up-front which isn't what I want.

I've also tested having the `query` use as similar `resolveField` method to `events`, but the issue persists.

**Comment by vladar** [comment]
We actually have a [bunch of tests](https://github.com/webonyx/graphql-php/blob/master/tests/Executor/ListsTest.php#L80) for deferred lists resolving.

I bet there is some bug in your code, but in order to help you find it, I need a reproducible test case or at least a repository where I can reproduce it.

Information in issue description is not enough to figure out what's wrong.

**Comment by Tam** [comment]
I've managed to get it _somewhat_ working. I was trying to return the ObjectType again when resolving instead of the Deferred.

I've now hit an issue where the query will stop at the first Deferred. If I run the following query, where `events` returns a Deferred:
```graphql
{ events { id handle } }
```
the query will stop at `events` and won't continue to process `id` or `handle`. Ideally I need the entire query to be processed before any deferred are resolved to allow me to build the SQL query.

I've created a gist with my current progress: https://gist.github.com/Tam/e76c64130b69ef31e19106f798434838.

**Comment by vladar** [comment]
But how it can process `id` or `handle` without the corresponding event data? I feel that you misunderstand deferreds, but not sure how to help you here.

**Comment by vladar** [comment]
One comment I have though is that building SQL for a whole query is usually a bad idea. The best performance achieved usually by using multiple simple (and properly indexed) queries and using IN(?) for batching. Such queries can also be cached easily (and benefit from [dataloader approach](https://github.com/facebook/dataloader#batching)).

Dynamic query building is very limited (especially if you have abstract types in your schema). I think you faced those limitations already. Also, performance won't be stable because it is hard to build indexes for such queries.

**Comment by mfn** [comment]
I was facing the same problem but used `\GraphQL\Type\Definition\ResolveInfo::getFieldSelection` to eager inspect which fields are going to be required (in fact I need a derivative version because I needn't just fields, also it's arguments).

I learned this works best on top of a specialized GraphQL query type which then, from the root, performs this inspection.

In my case I have a **very** complex SQL statement (nested with multiple sub-select, group by, window functions, lateral, etc.) on a table which has >90 columns (sorry, that sometimes is just the reality).

You would never load all fields in a batch, simply not practical for performance. Same as I don't expose all these fields anyway via GraphQL.

However I learned you will more easily run into "field dependencies": sometimes you code requires a field present if when not explicitly selected the user.

All in all it's horrible to solve this well, but it works and I just slapped enough unit tests onto it to make sure it doesn't break unnoticed 🤷‍♀️

I just wanted to give perspective of what we all strive to have and what we have to work with, in practice 😄 

The fact that it's solvable via the GraphQL library is also a testament how practically good the library is ;-)
